File: kmp-mobile/ios/PazChurch/Features/Profile/EditProfileViewModel.swift
import FirebaseStorage
import Observation
import Shared
import SwiftUI

@MainActor
@Observable
class EditProfileViewModel {
    var name = ""
    var phone = ""
    var birthDate: Date?
    var pictureUrl: String?
    var isUploadingPicture = false

    var cep = ""
    var street = ""
    var number = ""
    var complement = ""
    var neighborhood = ""
    var city = ""
    var state = ""
    var isLookingUpCep = false

    var isLoading = true
    var isSaving = false
    var error: String?
    var saveSuccess = false

    private let userRepository: UserRepository
    private let authRepository: AuthRepository
    private let onboardingRepository: OnboardingRepository

    private let isoFormatter: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withFullDate]
        return f
    }()

    init(
        userRepository: UserRepository,
        authRepository: AuthRepository,
        onboardingRepository: OnboardingRepository
    ) {
        self.userRepository = userRepository
        self.authRepository = authRepository
        self.onboardingRepository = onboardingRepository
        loadProfile()
    }

    // Sources truth from GET /users/me, NOT the cached login-time user: the
    // social-login response only ever carries id/name/email/picture/role, so
    // phone/birth_date/address are always absent from it — reading from the
    // cache here would make freshly-saved fields look like they never saved.
    private func loadProfile() {
        Task {
            do {
                let user = try await userRepository.getProfile()
                self.name = user.name
                self.phone = user.phone ?? ""
                self.pictureUrl = user.picture
                if let birthStr = user.birthDate {
                    self.birthDate = isoFormatter.date(from: birthStr)
                }
                if let address = user.addressDetails {
                    self.cep = address.zipCode ?? ""
                    self.street = address.street ?? ""
                    self.number = address.number ?? ""
                    self.complement = address.complement ?? ""
                    self.neighborhood = address.neighborhood ?? ""
                    self.city = address.city ?? ""
                    self.state = address.state ?? ""
                }
                self.isLoading = false
            } catch {
                self.error = error.localizedDescription
                self.isLoading = false
            }
        }
    }

    func onCepChanged(_ newValue: String) {
        let digits = String(newValue.filter(\.isNumber).prefix(8))
        cep = digits
        if digits.count == 8 { lookupCep(digits) }
    }

    private func lookupCep(_ cep: String) {
        Task {
            isLookingUpCep = true
            let outcome = try? await onboardingRepository.lookupCep(cep: cep)
            if let found = outcome as? CepLookupOutcomeFound {
                street = found.result.street
                neighborhood = found.result.neighborhood
                city = found.result.city
                state = found.result.state
            }
            isLookingUpCep = false
        }
    }

    func onPictureSelected(data: Data) {
        Task {
            isUploadingPicture = true
            error = nil
            do {
                let storage = Storage.storage()
                let ref = storage.reference().child("profile-pictures/\(UUID().uuidString).jpg")
                _ = try await ref.putDataAsync(data)
                let url = try await ref.downloadURL()
                pictureUrl = url.absoluteString
                isUploadingPicture = false
            } catch {
                self.error = "Erro ao enviar foto: \(error.localizedDescription)"
                isUploadingPicture = false
            }
        }
    }

    func onSave() {
        let trimmedName = name.trimmingCharacters(in: .whitespaces)
        if trimmedName.isEmpty {
            error = "Nome não pode ser vazio"
            return
        }

        let hasAddress = !street.trimmingCharacters(in: .whitespaces).isEmpty || !cep.isEmpty
        if hasAddress, street.trimmingCharacters(in: .whitespaces).isEmpty
            || number.trimmingCharacters(in: .whitespaces).isEmpty
            || neighborhood.trimmingCharacters(in: .whitespaces).isEmpty
            || city.trimmingCharacters(in: .whitespaces).isEmpty
            || state.trimmingCharacters(in: .whitespaces).isEmpty {
            error = "Preencha todos os campos obrigatórios do endereço, ou deixe todos em branco"
            return
        }

        isSaving = true
        error = nil

        let rawPhone = phone.isEmpty ? nil : phone.filter(\.isNumber)
        let birthStr = birthDate.map { isoFormatter.string(from: $0) }
        let address: AddressRequest? = hasAddress
            ? AddressRequest(
                street: street.trimmingCharacters(in: .whitespaces),
                number: number.trimmingCharacters(in: .whitespaces),
                complement: complement.trimmingCharacters(in: .whitespaces).isEmpty
                    ? nil : complement.trimmingCharacters(in: .whitespaces),
                neighborhood: neighborhood.trimmingCharacters(in: .whitespaces),
                city: city.trimmingCharacters(in: .whitespaces),
                state: state.trimmingCharacters(in: .whitespaces),
                zipCode: cep,
                country: "Brasil"
            )
            : nil

        Task {
            do {
                let request = UpdateProfileRequest(
                    name: trimmedName,
                    picture: pictureUrl,
                    phone: rawPhone,
                    birthDate: birthStr,
                    address: address
                )
                _ = try await userRepository.updateProfile(request: request)
                saveSuccess = true
                isSaving = false
            } catch {
                self.error = error.localizedDescription
                isSaving = false
            }
        }
    }
}

File: kmp-mobile/android/src/main/kotlin/br/church/paz/android/ui/features/profile/EditProfileViewModel.kt
package br.church.paz.android.ui.features.profile

import android.content.Context
import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import br.church.paz.shared.domain.model.AddressRequest
import br.church.paz.shared.domain.model.CepLookupOutcome
import br.church.paz.shared.domain.model.UpdateProfileRequest
import br.church.paz.shared.domain.repository.OnboardingRepository
import br.church.paz.shared.domain.repository.UserRepository
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.receiveAsFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.util.UUID

class EditProfileViewModel(
    private val userRepository: UserRepository,
    private val onboardingRepository: OnboardingRepository,
) : ViewModel() {
    private val _uiState = MutableStateFlow(EditProfileUiState(isLoading = true))
    val uiState: StateFlow<EditProfileUiState> = _uiState.asStateFlow()

    private val _effect = Channel<EditProfileEffect>(Channel.BUFFERED)
    val effect = _effect.receiveAsFlow()

    init {
        loadProfile()
    }

    // Sources truth from GET /users/me, NOT the cached login-time user: the
    // social-login response only ever carries id/name/email/picture/role, so
    // phone/birth_date/address are always absent from it — reading from the
    // cache here would make freshly-saved fields look like they never saved.
    private fun loadProfile() {
        viewModelScope.launch {
            runCatching { userRepository.getProfile() }
                .onSuccess { user ->
                    _uiState.update {
                        it.copy(
                            isLoading = false,
                            name = user.name,
                            phone = user.phone.orEmpty(),
                            birthDate = user.birthDate,
                            pictureUrl = user.picture,
                            street = user.addressDetails?.street.orEmpty(),
                            number = user.addressDetails?.number.orEmpty(),
                            complement = user.addressDetails?.complement.orEmpty(),
                            neighborhood = user.addressDetails?.neighborhood.orEmpty(),
                            city = user.addressDetails?.city.orEmpty(),
                            state = user.addressDetails?.state.orEmpty(),
                            zipCode = user.addressDetails?.zipCode.orEmpty(),
                        )
                    }
                }.onFailure { e ->
                    _uiState.update { it.copy(isLoading = false, error = e.message ?: "Erro ao carregar perfil") }
                }
        }
    }

    fun onNameChanged(v: String) = _uiState.update { it.copy(name = v, error = null) }

    fun onPhoneChanged(raw: String) = _uiState.update { it.copy(phone = formatPhoneDigits(raw.filter(Char::isDigit)), error = null) }

    fun onBirthDateChanged(isoDate: String) = _uiState.update { it.copy(birthDate = isoDate) }

    fun onStreetChanged(v: String) = _uiState.update { it.copy(street = v) }

    fun onNumberChanged(v: String) = _uiState.update { it.copy(number = v) }

    fun onComplementChanged(v: String) = _uiState.update { it.copy(complement = v) }

    fun onNeighborhoodChanged(v: String) = _uiState.update { it.copy(neighborhood = v) }

    fun onCityChanged(v: String) = _uiState.update { it.copy(city = v) }

    fun onStateChanged(v: String) = _uiState.update { it.copy(state = v) }

    fun onZipCodeChanged(v: String) {
        val digits = v.filter(Char::isDigit).take(8)
        _uiState.update { it.copy(zipCode = digits) }
        if (digits.length == 8) lookupCep(digits)
    }

    private fun lookupCep(cep: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLookingUpCep = true) }
            when (val outcome = onboardingRepository.lookupCep(cep)) {
                is CepLookupOutcome.Found ->
                    _uiState.update {
                        it.copy(
                            isLookingUpCep = false,
                            street = outcome.result.street,
                            neighborhood = outcome.result.neighborhood,
                            city = outcome.result.city,
                            state = outcome.result.state,
                        )
                    }
                else -> _uiState.update { it.copy(isLookingUpCep = false) }
            }
        }
    }

    fun onPictureSelected(
        context: Context,
        uri: Uri,
    ) {
        viewModelScope.launch {
            _uiState.update { it.copy(isUploadingPicture = true, error = null) }
            runCatching {
                val ref = FirebaseStorage.getInstance().reference.child("profile-pictures/${UUID.randomUUID()}.jpg")
                ref.putFile(uri).await()
                ref.downloadUrl.await().toString()
            }.onSuccess { url ->
                _uiState.update { it.copy(isUploadingPicture = false, pictureUrl = url) }
            }.onFailure { e ->
                _uiState.update { it.copy(isUploadingPicture = false, error = e.message ?: "Erro ao enviar foto") }
            }
        }
    }

    fun onSave() {
        val state = _uiState.value
        val name = state.name.trim()
        if (name.isEmpty()) {
            _uiState.update { it.copy(error = "Nome não pode ser vazio") }
            return
        }

        val hasAddress = state.street.isNotBlank() || state.zipCode.isNotBlank()
        if (hasAddress &&
            (
                state.street.isBlank() ||
                    state.number.isBlank() ||
                    state.neighborhood.isBlank() ||
                    state.city.isBlank() ||
                    state.state.isBlank()
            )
        ) {
            _uiState.update { it.copy(error = "Preencha todos os campos obrigatórios do endereço, ou deixe todos em branco") }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, error = null) }
            val request =
                UpdateProfileRequest(
                    name = name,
                    picture = state.pictureUrl,
                    phone = state.phone.filter(Char::isDigit).ifBlank { null },
                    birthDate = state.birthDate,
                    address =
                        if (hasAddress) {
                            AddressRequest(
                                street = state.street.trim(),
                                number = state.number.trim(),
                                complement = state.complement.trim().ifBlank { null },
                                neighborhood = state.neighborhood.trim(),
                                city = state.city.trim(),
                                state = state.state.trim(),
                                zipCode = state.zipCode,
                                country = "Brasil",
                            )
                        } else {
                            null
                        },
                )
            runCatching { userRepository.updateProfile(request) }
                .onSuccess {
                    _uiState.update { it.copy(isSaving = false, saveSuccess = true) }
                    _effect.send(EditProfileEffect.SaveSuccess)
                }.onFailure { e ->
                    _uiState.update {
                        it.copy(isSaving = false, error = e.message ?: "Erro ao salvar perfil")
                    }
                }
        }
    }

    fun onBack() {
        viewModelScope.launch { _effect.send(EditProfileEffect.NavigateBack) }
    }
}

private fun formatPhoneDigits(digits: String): String {
    val d = digits.take(11)
    val sb = StringBuilder()
    d.forEachIndexed { i, c ->
        when (i) {
            0 -> sb.append('(').append(c)
            1 -> sb.append(c).append(") ")
            2 -> sb.append(c).append(' ')
            6 -> sb.append(c).append('-')
            else -> sb.append(c)
        }
    }
    return sb.toString()
}
